File: ios/LocalPods/ScannerModule/Core/Network/Models/FarmModels.swift
// LocalPods/ScannerModule/Core/Network/Models/FarmModels.swift
//
// Farm API models (§4)

import Foundation

// MARK: - Farm Response

struct Farm: Codable {
    let farmId: String
    let ownerDid: String
    let regionCode: String
    let farmName: String?
    let boundary: GeoJSONPolygon
    let lonOrigin: Double?
    let latOrigin: Double?
    let rowSpacing: Double?
    let colSpacing: Double?
    let createdAt: String
    let updatedAt: String

    enum CodingKeys: String, CodingKey {
        case farmId = "farm_id"
        case ownerDid = "owner_did"
        case regionCode = "region_code"
        case farmName = "farm_name"
        case boundary
        case lonOrigin = "lon_origin"
        case latOrigin = "lat_origin"
        case rowSpacing = "row_spacing"
        case colSpacing = "col_spacing"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

// MARK: - GeoJSON Polygon

struct GeoJSONPolygon: Codable {
    let type: String
    let coordinates: [[[Double]]] // [[[lon, lat]]]

    init(coordinates: [[[Double]]]) {
        self.type = "Polygon"
        self.coordinates = coordinates
    }
}

// MARK: - Farm List Response

struct FarmListData: Codable {
    let items: [Farm]
    let limit: Int
    let offset: Int
}

// MARK: - Farm Create Request (§4.2)

struct FarmCreateRequest: Codable {
    let farmId: String
    let ownerDid: String
    let regionCode: String
    let farmName: String?
    let boundary: GeoJSONPolygon

    enum CodingKeys: String, CodingKey {
        case farmId = "farm_id"
        case ownerDid = "owner_did"
        case regionCode = "region_code"
        case farmName = "farm_name"
        case boundary
    }
}

// MARK: - Farm Update Request (§4.4)

struct FarmUpdateRequest: Codable {
    let ownerDid: String
    let regionCode: String
    let farmName: String?
    let boundary: GeoJSONPolygon

    enum CodingKeys: String, CodingKey {
        case ownerDid = "owner_did"
        case regionCode = "region_code"
        case farmName = "farm_name"
        case boundary
    }
}

// MARK: - Farm Patch Request (§4.5)

struct FarmPatchRequest: Codable {
    let ownerDid: String?
    let regionCode: String?
    let farmName: String?
    let boundary: GeoJSONPolygon?

    enum CodingKeys: String, CodingKey {
        case ownerDid = "owner_did"
        case regionCode = "region_code"
        case farmName = "farm_name"
        case boundary
    }
}

// MARK: - Farm Grid Config (§4.8, §4.9)

struct FarmGridConfig: Codable {
    let farmId: String
    let totalTrees: Int
    let rowMin: Int?
    let rowMax: Int?
    let colMin: Int?
    let colMax: Int?
    let lonOrigin: Double?
    let latOrigin: Double?
    let rowSpacing: Double?
    let colSpacing: Double?

    enum CodingKeys: String, CodingKey {
        case farmId = "farm_id"
        case totalTrees = "total_trees"
        case rowMin = "row_min"
        case rowMax = "row_max"
        case colMin = "col_min"
        case colMax = "col_max"
        case lonOrigin = "lon_origin"
        case latOrigin = "lat_origin"
        case rowSpacing = "row_spacing"
        case colSpacing = "col_spacing"
    }
}

struct FarmGridConfigRequest: Codable {
    let lonOrigin: Double
    let latOrigin: Double
    let rowSpacing: Double
    let colSpacing: Double

    enum CodingKeys: String, CodingKey {
        case lonOrigin = "lon_origin"
        case latOrigin = "lat_origin"
        case rowSpacing = "row_spacing"
        case colSpacing = "col_spacing"
    }
}

// MARK: - Farm Tree List (§4.7)

struct FarmTreeListData: Codable {
    let items: [Tree]
    let total: Int
    let limit: Int
    let offset: Int
}

File: android/orilifesdk/src/main/java/com/mvp/orilife/network/models/FarmModels.kt
package com.mvp.orilife.network.models

import com.google.gson.annotations.SerializedName

/**
 * GeoJSON Polygon
 * Format: {"type":"Polygon","coordinates":[[[lng,lat],[lng,lat],...]]}
 */
data class GeoJsonPolygon(
    @SerializedName("type")
    val type: String = "Polygon",

    @SerializedName("coordinates")
    val coordinates: List<List<List<Double>>>  // [[[lng,lat],[lng,lat],...]]
)

/**
 * Farm response from backend (§4.2)
 */
data class Farm(
    @SerializedName("farm_id")
    val farmId: String,

    @SerializedName("owner_did")
    val ownerDid: String,

    @SerializedName("region_code")
    val regionCode: String,

    @SerializedName("farm_name")
    val farmName: String?,

    @SerializedName("boundary")
    val boundary: GeoJsonPolygon,

    @SerializedName("lon_origin")
    val lonOrigin: Double?,

    @SerializedName("lat_origin")
    val latOrigin: Double?,

    @SerializedName("row_spacing")
    val rowSpacing: Double?,

    @SerializedName("col_spacing")
    val colSpacing: Double?,

    @SerializedName("created_at")
    val createdAt: String,

    @SerializedName("updated_at")
    val updatedAt: String
)

/**
 * Farm create request (§4.2)
 */
data class FarmCreateRequest(
    @SerializedName("farm_id")
    val farmId: String,

    @SerializedName("owner_did")
    val ownerDid: String,

    @SerializedName("region_code")
    val regionCode: String,

    @SerializedName("farm_name")
    val farmName: String?,

    @SerializedName("boundary")
    val boundary: GeoJsonPolygon
)

/**
 * Grid config request (§4.9)
 */
data class GridConfigRequest(
    @SerializedName("lon_origin")
    val lonOrigin: Double,

    @SerializedName("lat_origin")
    val latOrigin: Double,

    @SerializedName("row_spacing")
    val rowSpacing: Double,

    @SerializedName("col_spacing")
    val colSpacing: Double
)

/**
 * Grid config response (§4.8)
 */
data class GridConfigResponse(
    @SerializedName("farm_id")
    val farmId: String,

    @SerializedName("total_trees")
    val totalTrees: Int,

    @SerializedName("row_min")
    val rowMin: Int?,

    @SerializedName("row_max")
    val rowMax: Int?,

    @SerializedName("col_min")
    val colMin: Int?,

    @SerializedName("col_max")
    val colMax: Int?,

    @SerializedName("lon_origin")
    val lonOrigin: Double?,

    @SerializedName("lat_origin")
    val latOrigin: Double?,

    @SerializedName("row_spacing")
    val rowSpacing: Double?,

    @SerializedName("col_spacing")
    val colSpacing: Double?
)

/**
 * Farm list response (§4.1)
 */
data class FarmListData(
    @SerializedName("items")
    val items: List<Farm>,

    @SerializedName("limit")
    val limit: Int,

    @SerializedName("offset")
    val offset: Int
)
